### PR TITLE
Add Multi-threaded Batch Document

### PR DESCRIPTION
**Pull request by HAKSOAT**
This addresses #1778

### CODE REVIEW

**Comment by HAKSOAT** [comment]
Hi, @lintool I have a PR in progress. I need your thoughts on some things:

1. The original `document` methods [L663](https://github.com/castorini/anserini/blob/b429218e52a385eabf3fd81979e221111fbc4a19/src/main/java/io/anserini/search/SimpleSearcher.java#L663) and [L680](https://github.com/castorini/anserini/blob/b429218e52a385eabf3fd81979e221111fbc4a19/src/main/java/io/anserini/search/SimpleSearcher.java#L680) take in an integer and a string respectively. However, I'm not very sure how I can go about doing this without creating two `batchDocument` methods with different names. Hence, I'd like to know what the best practice is for Anserini. The PR only currently tackles batch for internal lucene docs and I'd like to add something for collection docid too.

2. After compiling I saw a bunch of modified markdown files. Will I push all of them as well? I didn't push because I didn't want them to mess up the PR review.

**Comment by lintool** [comment]
hi @HAKSOAT - thanks for this PR.

Let's go with a single method for now, called `batchGetDocument`. It should take a list of (external) collection ids, which is the more common case.

If you update and merge in current master, it should get rid of the new markdown edits?

Thanks!

File: src/test/java/io/anserini/search/SimpleSearcherTest.java [hunk 1]
```diff
@@ -54,6 +55,29 @@ public void testGetDoc() throws Exception {
...
+
+    ArrayList<Integer> ldocIds = new ArrayList<Integer>();
+    ldocIds.add(0);
+    ldocIds.add(1);
+    ldocIds.add(2);
+    ldocIds.add(3);
+
+    Map<Integer, Document> results = searcher.batchDocument(ldocIds, 2);
+
+    assertEquals("here is some text here is some more text. city.",
```
Comment (lintool) [inline src/test/java/io/anserini/search/SimpleSearcherTest.java:70]: I prefer _not_ to break these lines...

File: src/main/java/io/anserini/search/SimpleSearcher.java [hunk 2]
```diff
@@ -681,6 +681,45 @@ public Document document(String docid) {
...
   }
 
+  /**
+   * Returns a map of collection docid to Lucene {@link Document}.
+   * Batch version of {@link #document(String)}.
+   *
+   * @param docids list of docids
+   * @return a map of docid to corresponding Lucene {@link Document}
+   */
+  public Map<String, Document> batchGetDocument(List<String> docids, int threads){
```
Comment (lintool) [inline src/main/java/io/anserini/search/SimpleSearcher.java:691]: space before { here.

**Comment by lintool** [comment]
lg, but can you please build a fatjar, copy over to pyserini, and make sure it connects up on the pyserini end? i.e., let's build the bindings on the pyserini end just to confirm everything works e2e.

**Comment by HAKSOAT** [comment]
> lg, but can you please build a fatjar, copy over to pyserini, and make sure it connects up on the pyserini end? i.e., let's build the bindings on the pyserini end just to confirm everything works e2e.

Noted.

**Comment by HAKSOAT** [comment]
> lg, but can you please build a fatjar, copy over to pyserini, and make sure it connects up on the pyserini end? i.e., let's build the bindings on the pyserini end just to confirm everything works e2e.

It works. I'll put a PR through for this on Pyserini.

**Review by lintool — APPROVED** [review]

**Pull request merged**